src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcCoreDiagnosticSourceExtensions.cs

```diff
@@ -16,55 +16,82 @@ namespace Microsoft.AspNetCore.Mvc.Internal
     public static class MvcCoreDiagnosticSourceExtensions
     {
         public static void BeforeAction(
-            this DiagnosticSource diagnosticSource,
+            this DiagnosticListener diagnosticListener,
             ActionDescriptor actionDescriptor,
             HttpContext httpContext,
             RouteData routeData)
         {
-            Debug.Assert(diagnosticSource != null);
+            Debug.Assert(diagnosticListener != null);
             Debug.Assert(actionDescriptor != null);
             Debug.Assert(httpContext != null);
             Debug.Assert(routeData != null);
 
-            if (diagnosticSource.IsEnabled("Microsoft.AspNetCore.Mvc.BeforeAction"))
+            // Inlinable fast-path check if Diagnositcs is enabled
+            if (diagnosticListener.IsEnabled())
             {
-                diagnosticSource.Write(
+                BeforeActionImpl(diagnosticListener, actionDescriptor, httpContext, routeData);
+            }
+        }
+
+        private static void BeforeActionImpl(DiagnosticListener diagnosticListener, ActionDescriptor actionDescriptor, HttpContext httpContext, RouteData routeData)
+        {
+            if (diagnosticListener.IsEnabled("Microsoft.AspNetCore.Mvc.BeforeAction"))
```

Comment:
Is there any value to having this check appear again?

Reply:
Oh wait I see, you're checking if the listener is enabled at all before checking if the specific category is enabled. I can read 😆

Reply:
Fast inlining check just says if any loggers are active. Second virtual check is to see if any is listening to that event